File: dotnet/src/Wata.Sdk/Errors/WataApiException.cs
using System.Text.Json;

namespace Wata.Sdk.Errors;

/// <summary>
/// 4xx с телом { error: { code, message, details, validationErrors } }.
/// Содержит код ошибки WATA, сообщение и HTTP-статус.
/// </summary>
public sealed class WataApiException : WataException
{
    public WataApiException(string message, string? code, string? details, JsonElement? validationErrors)
        : base(message)
    {
        Code = code;
        Details = details;
        ValidationErrors = validationErrors;
    }

    /// <summary>Код ошибки WATA, например PL_NOT_FOUND, TRA_ALREADY_REFUNDED и т.п.</summary>
    public string? Code { get; }

    /// <summary>Дополнительные детали ошибки от сервера, если есть.</summary>
    public string? Details { get; }

    /// <summary>
    /// Сырое содержимое поля validationErrors — формат не зафиксирован спецификацией,
    /// поэтому оно сохраняется как JsonElement, чтобы не терять данные при неожиданной форме.
    /// </summary>
    public JsonElement? ValidationErrors { get; }
}

File: java/src/main/java/pro/wata/sdk/errors/WataApiException.java
package pro.wata.sdk.errors;

import java.util.List;
import java.util.Map;

/**
 * Ошибка {@code 4xx} (кроме 401/403/429) с телом
 * {@code {error: {code, message, details, validationErrors}}}. Несёт код
 * ошибки WATA ({@code PL_*}, {@code TRA_*}, {@code ORD_*}, {@code STM_*},
 * {@code STR_*}, {@code TPP_*}, {@code VCR_*}), сообщение и HTTP-статус.
 */
public class WataApiException extends WataException {

    private final Object details;
    private final List<Map<String, Object>> validationErrors;

    public WataApiException(String message, int httpStatus, String requestPath, String wataErrorCode,
                             Object details, List<Map<String, Object>> validationErrors) {
        super(message, httpStatus, requestPath, wataErrorCode);
        this.details = details;
        this.validationErrors = validationErrors;
    }

    /** Дополнительные детали ошибки, если сервер их прислал. */
    public Object details() {
        return details;
    }

    /** Ошибки валидации отдельных полей запроса, если применимо. */
    public List<Map<String, Object>> validationErrors() {
        return validationErrors;
    }
}
